### PR TITLE
Add runtime options API for ORTModule

### PR DESCRIPTION
**Pull request by adk9**
**Description**: Add runtime options API for ORTModule

**Motivation and Context**
- Setting any runtime options in ORTModule currently requires either using an internal API or using undocumented environment variables. This PR introduces a `runtime_options` argument to `ORTModule` which can be used to set several options related to the ORTModule (torch-onnx) exporter or the runtime session. This is similar to the corresponding `debug_options` argument currently present to set debug options in ORTModule.

The available runtime options at the moment are:
- **onnx_opset_version** (`int`, optional, default 14): Configure ORTModule exporter to use this ONNX opset version.
            To change the opset version, the environment variable "ORTMODULE_ONNX_OPSET_VERSION" can also be used.
- **do_constant_folding** (`bool`, optional): Configure ORTModule export to apply the constant-folding optimization.
            Defaults to False. Constant-folding will replace some of the ops that have all constant inputs with pre-computed
            constant nodes.
- **export_modules_as_functions** (`bool` or set of python:type of nn.Module, optional, default False): ORTModule exporter
            flag to enable exporting all nn.Module forward calls as local functions in ONNX. Or a set to indicate the particular
            types of modules to export as local functions in ONNX. This feature requires onnx_opset_version >= 15, otherwise the
            export will fail. This is because onnx_opset_version < 15 implies IR version < 8, which means no local function support.
- **export_params** (`bool`, optional, default False): ORTModule exporter flag to export all model parameters.
            If True, all parameters will be exported. Set this to False if you want to export an untrained model. In this case,
            the exported model will first take all of its parameters as arguments.
- **keep_initializers_as_inputs** (`bool`, optional, default True): ORTModule exporter flag to keep initializers as inputs.
            If True, all the initializers (typically corresponding to parameters) in the exported graph will also be added as inputs
            to the graph. If False, initializers are not added as inputs to the graph, and only the non-parameter inputs are added as inputs.
            This may allow for better optimizations (e.g. constant folding) by backends/runtimes.
- **use_static_shapes** (`bool`, optional, default False): ORTModule exporter flag to use static output shapes.
            By default the exported model will have dynamic shapes for all input and output tensors.
- **exporter_extra_args** (`dict`, optional, default Empty): Extra arguments to specify to the ORTModule Torch ONNX exporter.
            Refer to https://pytorch.org/docs/stable/onnx.html#torch.onnx.export for documentation on exporter arguments.
- **enable_custom_autograd** (:obj:`bool`, optional, default False): Enable custom autograd.Function support for ORTModule.
            This will enable ``autograd.Function``s to be exported as ``PythonOp`` in ONNX.
- **disable_custom_ops** (`bool`, optional, default False): Disable custom ops support in ORTModule.
            This flag will disable custom ATen ops and custom gradients in ORTModule.
* **graph_optimization_level** (`onnxruntime.GraphOptimizationLevel`, optional, default None): Choose the graph optimization
            level for ORTModule runtime session. The available graph optimization levels are:
  - ``onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL``: Disable all optimizations
  - ``onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC``: Constant folding and other optimizations that only use ONNX operators
  - ``onnxruntime.GraphOptimizationLevel.ORT_ENABLE_EXTENDED``: Optimizations using custom operators, excluding NCHWc and NHWC layout optimizers
  - ``onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL``: Enable all optimizations (default)
 
### Raises:
- **TypeError**: If an invalid type for a runtime option is specified.
- **ValueError**: If an invalid runtime option value is specified.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/runtime_options.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        use_static_shapes (:obj:`bool`, optional, default False): ORTModule exporter flag to use static output shapes.
+            By default the exported model will have dynamic shapes for all input and output tensors.
+
+        exporter_extra_args (:obj:`dict`, optional, default Empty): Extra arguments to specify to the ORTModule Torch ONNX exporter.
+            Refer to https://pytorch.org/docs/stable/onnx.html#torch.onnx.export for documentation on exporter arguments.
+
+        enable_custom_autograd (:obj:`bool`, optional, default False): Enable custom autograd.Function support for ORTModule.
+            This will enable ``autograd.Function``s to be exported as ``PythonOp`` in ONNX.
+
+        disable_custom_ops (:obj:`bool`, optional, default False): Disable custom ops support in ORTModule.
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/runtime_options.py:147]: what is the intention of this option? Do you want to disable aten op overrides or aten ops altogether? 

Some Context: Today the only way to enable or disable aten ops is via the symbolic registry... soon we will enable the ATEN_FALLBACK option in exporter (which will be available starting pytorch 1.12)... with this change, when this option is selected any op which cannot be represented with a single onnx op or a combination of onnx ops will be exported as ATEN op. 

Going back to my question - do you want to disable just the overrides or turn off aten op export?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -412,13 +419,24 @@ def _get_exported_model(self, input_schema, *inputs, **kwargs):
...
-                    "dynamic_axes": self._input_info.dynamic_axes,
                     "verbose": self._debug_options.logging.log_level < LogLevel.WARNING,
-                    "export_params": False,
-                    "keep_initializers_as_inputs": True,
+                    "export_params": self._runtime_options.exporter_options.export_params,
+                    "keep_initializers_as_inputs": self._runtime_options.exporter_options.keep_initializers_as_inputs,
                 }
+
+                if not self._runtime_options.exporter_options.use_static_shapes:
+                    required_export_kwargs["dynamic_axes"] = self._input_info.dynamic_axes
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:430]: nit: you can assign the value on line 417 itself no need to add this here
some_expression if condition else other_expression

**Comment by thiagocrepaldi** [comment]
All those settings are meant to be hidden/private by design
PyTorch users dont need to understand implementation details of ONNX exporter. Instead, ORTModule should be able to set the best settings on its own. When it is not possible to do so, ORTModule can be extended (inheritance) to make user-specific changes or the internals can be used by whom really know what they are doing

Exposing internal complexity will only make difficult for users and make it similar to the deprecated `ORTTrainer` frontend

File: orttraining/orttraining/python/training/ortmodule/runtime_options.py (hunk 1 above)
Comment (adk9) [reply orttraining/orttraining/python/training/ortmodule/runtime_options.py:147]: The intention behind introducing this option is to be able to turn off ATen op export in the graph as external/custom EPs may not fully support them. Is there a way to disable them in the exporter?

File: orttraining/orttraining/python/training/ortmodule/runtime_options.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            This flag will disable custom ATen ops and custom gradients in ORTModule.
+
+        graph_optimization_level (:obj:`onnxruntime.GraphOptimizationLevel`, optional, default None): Choose the graph optimization
```
[the remark this answers is not in the document]
Comment (adk9) [reply orttraining/orttraining/python/training/ortmodule/runtime_options.py:150]: The use case here is an external JIT compilation pipeline that requires the unoptimized graph for better control over which optimizations to apply.

**Comment by baijumeswani** [comment]
I second @thiagocrepaldi's concerns here. Are all these options needed? `ORTModule` was designed to look and feel like a pytorch model hiding all the behind-the-scenes implementation away from the users. Users should not need to know if export is required or how `ORTModule` performs execution.

Are all the options mentioned in the description required? Is there a way we can avoid passing these settings into the constructor? Can these settings be controlled through other means? Would you elaborate what specific use case these flags help in resolving?

**Comment by ytaous** [comment]
please resolve conflicts.

**Comment by pengwa** [comment]
> 
OK, I understand the intention now. 

One use case could be like this: we have few perf related options, not one default value can serve all models well, sometimes, when we on board models, we can potentially tune the perf using different values, choose a best one fitting that model. How we allow this dynamic config, with environment variable? that is actually what I have to use in one of previous trials. https://github.com/microsoft/onnxruntime/pull/11359/files#diff-117ceac0be922292ecb1559bbd892ae69e6882ab1e0a35823cf24539dd17b2cfR153 

Any suggestions for this use case?
